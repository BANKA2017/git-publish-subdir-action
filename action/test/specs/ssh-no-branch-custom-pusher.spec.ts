import * as path from 'path';

import * as util from '../util';

it('Custom Pusher', async () => {

  const REPO_DIR = path.join(util.REPOS_DIR, 'ssh-no-branch-custom-pusher.git');
  const DATA_DIR = path.join(util.DATA_DIR, 'ssh-no-branch-custom-pusher');

  // Create empty repo
  await util.mkdir(REPO_DIR);
  await util.execWithOutput('git init --bare', { cwd: REPO_DIR });

  // Create dummy data
  await util.mkdir(DATA_DIR);
  await util.mkdir(path.join(DATA_DIR, 'dummy'));
  await util.writeFile(path.join(DATA_DIR, 'dummy', 'baz'), 'foobar');
  await util.writeFile(path.join(DATA_DIR, 'dummy', '.bat'), 'foobar');

  // Run Action
  await util.runWithGithubEnv(
    path.basename(__filename),
    {
      REPO: 'ssh://git@git-ssh/git-server/repos/ssh-no-branch-custom-pusher.git',
      BRANCH: 'branch-a',
      FOLDER: DATA_DIR,
      SSH_PRIVATE_KEY: (await util.readFile(util.SSH_PRIVATE_KEY)).toString(),
      KNOWN_HOSTS_FILE: util.KNOWN_HOSTS,
    },
    's0/test',
    {
      pusher: {
        email: 'bob@examle.com',
        name: 'Alice Bob',
      }
    },
    's0',
  );

  // Check that the log of the repo is as expected
  // (check tree-hash, commit message, and author)
  const log = (await util.exec(
    'git log --pretty="format:msg:%s%ntree:%T%nauthor:%an <%ae>" branch-a',
    {
      cwd: REPO_DIR
    }
  )).stdout;
  const sha = await util.getRepoSha();
  const cleanedLog = log.replace(sha, '<sha>');
  expect(cleanedLog).toMatchSnapshot();
});

it('Custom Pusher (invalid)', async () => {

  const REPO_DIR = path.join(util.REPOS_DIR, 'ssh-no-branch-custom-pusher-invalid.git');
  const DATA_DIR = path.join(util.DATA_DIR, 'ssh-no-branch-custom-pusher-invalid');

  // Create empty repo
  await util.mkdir(REPO_DIR);
  await util.execWithOutput('git init --bare', { cwd: REPO_DIR });

  // Create dummy data
  await util.mkdir(DATA_DIR);
  await util.mkdir(path.join(DATA_DIR, 'dummy'));
  await util.writeFile(path.join(DATA_DIR, 'dummy', 'baz'), 'foobar');
  await util.writeFile(path.join(DATA_DIR, 'dummy', '.bat'), 'foobar');

  // Run Action
  await util.runWithGithubEnv(
    path.basename(__filename),
    {
      REPO: 'ssh://git@git-ssh/git-server/repos/ssh-no-branch-custom-pusher-invalid.git',
      BRANCH: 'branch-a',
      FOLDER: DATA_DIR,
      SSH_PRIVATE_KEY: (await util.readFile(util.SSH_PRIVATE_KEY)).toString(),
      KNOWN_HOSTS_FILE: util.KNOWN_HOSTS,
    },
    's0/test',
    {
      pusher: {}
    },
    's0',
  );

  // Check that the log of the repo is as expected
  // (check tree-hash, commit message, and author)
  const log = (await util.exec(
    'git log --pretty="format:msg:%s%ntree:%T%nauthor:%an <%ae>" branch-a',
    {
      cwd: REPO_DIR
    }
  )).stdout;
  const sha = await util.getRepoSha();
  const cleanedLog = log.replace(sha, '<sha>');
  expect(cleanedLog).toMatchSnapshot();
});

it('No Pusher or Actor', async () => {

  const REPO_DIR = path.join(util.REPOS_DIR, 'ssh-no-branch-custom-pusher-none.git');
  const DATA_DIR = path.join(util.DATA_DIR, 'ssh-no-branch-custom-pusher-none');

  // Create empty repo
  await util.mkdir(REPO_DIR);
  await util.execWithOutput('git init --bare', { cwd: REPO_DIR });

  // Create dummy data
  await util.mkdir(DATA_DIR);
  await util.mkdir(path.join(DATA_DIR, 'dummy'));
  await util.writeFile(path.join(DATA_DIR, 'dummy', 'baz'), 'foobar');
  await util.writeFile(path.join(DATA_DIR, 'dummy', '.bat'), 'foobar');

  // Run Action
  await util.runWithGithubEnv(
    path.basename(__filename),
    {
      REPO: 'ssh://git@git-ssh/git-server/repos/ssh-no-branch-custom-pusher-none.git',
      BRANCH: 'branch-a',
      FOLDER: DATA_DIR,
      SSH_PRIVATE_KEY: (await util.readFile(util.SSH_PRIVATE_KEY)).toString(),
      KNOWN_HOSTS_FILE: util.KNOWN_HOSTS,
    },
    's0/test',
  );

  // Check that the log of the repo is as expected
  // (check tree-hash, commit message, and author)
  const log = (await util.exec(
    'git log --pretty="format:msg:%s%ntree:%T%nauthor:%an <%ae>" branch-a',
    {
      cwd: REPO_DIR
    }
  )).stdout;
  const sha = await util.getRepoSha();
  const cleanedLog = log.replace(sha, '<sha>');
  expect(cleanedLog).toMatchSnapshot();
});