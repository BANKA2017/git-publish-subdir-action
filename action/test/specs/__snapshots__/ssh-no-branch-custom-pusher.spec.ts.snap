// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Pusher (invalid) 1`] = `
"msg:Update branch-a to output generated at <sha>
tree:8bf87c66655949e66937b11593cc4ae732d1f610
author:s0 <s0@users.noreply.github.com>
msg:Initial branch-a commit
tree:f93e3a1a1525fb5b91020da86e44810c87a2d7bc
author:s0 <s0@users.noreply.github.com>"
`;

exports[`Custom Pusher 1`] = `
"msg:Update branch-a to output generated at <sha>
tree:8bf87c66655949e66937b11593cc4ae732d1f610
author:Alice Bob <bob@examle.com>
msg:Initial branch-a commit
tree:f93e3a1a1525fb5b91020da86e44810c87a2d7bc
author:Alice Bob <bob@examle.com>"
`;

exports[`No Pusher or Actor 1`] = `
"msg:Update branch-a to output generated at <sha>
tree:8bf87c66655949e66937b11593cc4ae732d1f610
author:Git Publish Subdirectory <nobody@nowhere>
msg:Initial branch-a commit
tree:f93e3a1a1525fb5b91020da86e44810c87a2d7bc
author:Git Publish Subdirectory <nobody@nowhere>"
`;
